feat(http server): add new builder APIs `build_from_tcp` and `build_from_hyper`

niklasad1 opened the pull request:
Before the HTTP server configured the socket explictly but this PR changes to use `tokio::net::TcpListener` similar to what the `WsServer does`

In addition this PR adds another API for users to provide their own TcpListener, see the inline example I added with socket2.

Example:


 ```rust
	/// use jsonrpsee_http_server::HttpServerBuilder;
	/// use socket2::{Domain, Socket, Type};
	///
	/// #[tokio::main]
	/// async fn main() {
	///   let addr = "127.0.0.1:0".parse().unwrap();
	///   let domain = Domain::for_address(addr);
	///   let socket = Socket::new(domain, Type::STREAM, None).unwrap();
	///   socket.set_nodelay(true).unwrap();
	///   socket.set_reuse_address(true).unwrap();
	///   socket.set_nonblocking(true).unwrap();
	///   socket.set_keepalive(true).unwrap();
	///
	///   let address = addr.into();
	///   socket.bind(&address).unwrap();
	///
	///   socket.listen(4096).unwrap();
	///
	///   let server = HttpServerBuilder::new().build_from_tcp(socket).await.unwrap();
	/// }
	/// ```

```

niklasad1 committed:
unify servers; use tokio::TcpListener

niklasad1 committed:
add API to build server from tcp socket

niklasad1 committed:
Merge remote-tracking branch 'origin/master' into na-tcp-socket-configurations-api

niklasad1 committed:
build_from_tcp nit: not async

niklasad1 committed:
Merge remote-tracking branch 'origin/master' into na-tcp-socket-configurations-api

jsdw commented:
Ooh offhand this looks very neat, I like it! I also like that it removes socket2 as a dependency and leaves it to the user to provide their own socket thing (if I read it right)!

niklasad1 commented:
yeah, you are right and that is the idea :)

AFAIU, most socket libraries implements `Into<StdTcpListener>` and `hyper` has a similar API

niklasad1 committed:
fix nits

jsdw commented on http-server/src/server.rs:
```diff
@@ -178,56 +218,26 @@ impl<M> Builder<M> {
 		let local_addr = listener.local_addr().ok();
-		let listener = hyper::Server::from_tcp(listener)?;
-		Ok((listener, local_addr))
+		let listener = hyper::Server::from_tcp(listener)?.tcp_nodelay(true);
```
in the comment above, you say that hyper sets tcp_nodelay == true, so is this `nodelay(true)` going to be ignored anyway?

niklasad1 replied:
yeah, I was thinking that we might want to expose these settings in the builder.

It's like a warning that even if folk configure these on their own sockets it might get overwritten but let's revert this then.

niklasad1 replied:
nah, my comment was wrong hyper sets it to false

jsdw reviewed (approved):
Loooks good to me; I really like letting somebody provide (and configure) their own socket rather than handling it internally :)

niklasad1 committed:
change `build_from_tcp` force users configure hyper

The hyper settings might contradict to settings on the provided
socket, force users of this API to configure that avoid confusion and
unexpected settings.

niklasad1 commented on http-server/src/server.rs:
```diff
@@ -167,6 +157,56 @@ impl<M> Builder<M> {
+	///
+	///   socket.listen(4096).unwrap();
+	///
+	///   // hyper does some settings on the provided socket, ensure that nothing breaks the our settings.
```
```suggestion
	///   // hyper does some settings on the provided socket, ensure that nothing breaks our "expected settings".
```

niklasad1 committed:
Update http-server/src/server.rs

niklasad1 commented on http-server/src/server.rs:
````diff
@@ -167,6 +157,56 @@ impl<M> Builder<M> {
+	///   let server = HttpServerBuilder::new().build_from_tcp(socket, hyper_cfg).unwrap();
+	/// }
+	/// ```
+	pub fn build_from_tcp(self, listener: impl Into<StdTcpListener>, cfg: HyperTcpConfig) -> Result<Server<M>, Error> {
````
@jsdw is this API reasonable?

Basically the only thing I want to ensure is that no settings gets overwritten without anything informing the users of this API.

It's a bit awkward to some settings twice but no way around that....

niklasad1 replied:
fair enough, I think your suggestion is less awkward.

niklasad1 committed:
fix doc links

niklasad1 committed:
Merge remote-tracking branch 'origin/na-tcp-socket-configurations-api' into na-tcp-socket-configurations-api

niklasad1 committed:
add separate APIs `build_from_tcp` and `build_from_hyper`

niklasad1 committed:
Merge remote-tracking branch 'origin/master' into na-tcp-socket-configurations-api

niklasad1 committed:
fix clippy

niklasad1 committed:
remove dead code

niklasad1 committed:
fix build

niklasad1 committed:
fix nit in example code
